### PR TITLE
feat(backend): 교수 등록과 호감도 코어 구현

### LINKED ISSUE #4: Task 1.4 교수, 호감도, 등록 API 코어
## Sprint
- Sprint 1: 기반 고정과 US1 병렬 구현

## 분류
- 작업 번호: Task 1.4
- 사용자 스토리: US1
- GitHub 라벨: Back-end, Feature

## 담당
- 개발자 A

## 목표
- 프로젝트 골격, 공통 기반, 인증, 교수 등록/조회, 캐릭터 에셋 상태 조회까지 완료한다.

## 독립 검증 기준
- 신규 사용자가 회원 가입 후 로그인하고 교수 1명을 등록한 뒤, 교수 상세에서 `characterAssetStatus`, 기본 호감도, 대표 에셋 또는 fallback 상태를 확인할 수 있어야 한다.

## 의존성
- 선행 이슈: #2 Task 1.2 공통 런타임과 테스트 기반 구축
- 병렬 가능 이슈: #5 Task 1.5 캐릭터 에셋 생성 파이프라인

## Atomic Tasks
- [ ] T017 [P] [US1] `Backend/src/test/java/com/animalleague/april/contract/professor/ProfessorContractTest.java`에 교수 등록/목록/상세 계약과 `male`, `female` 성별 검증 실패 테스트를 추가한다.
- [ ] T018 [P] [US1] `Backend/src/test/java/com/animalleague/april/integration/professor/ProfessorRegistrationIntegrationTest.java`에 교수 등록과 기본 호감도 생성 통합 실패 테스트를 추가한다.
- [ ] T019 [P] [US1] `Backend/src/test/java/com/animalleague/april/unit/professor/AffectionInitializationUnitTest.java`에 교수 등록 시 호감도 초기화 단위 실패 테스트를 추가한다.
- [ ] T020 [US1] `Backend/src/main/resources/db/migration/V2__create_professor_affection_tables.sql`에 `professors`, `affections` 테이블 스키마를 정의한다.
- [ ] T021 [US1] `Backend/src/main/java/com/animalleague/april/professor/domain/Professor.java`와 `Backend/src/main/java/com/animalleague/april/professor/domain/Affection.java`에 교수와 호감도 애그리거트를 구현한다.
- [ ] T022 [US1] `Backend/src/main/java/com/animalleague/april/professor/infrastructure/ProfessorRepository.java`, `Backend/src/main/java/com/animalleague/april/professor/infrastructure/AffectionRepository.java`, `Backend/src/main/java/com/animalleague/april/professor/application/ProfessorService.java`, `Backend/src/main/java/com/animalleague/april/professor/api/ProfessorController.java`에 교수 등록/조회 코어를 구현한다.

## 참고 문서
- `specs/001-backend-mvp/tasks.md`
- `specs/001-backend-mvp/spec.md`
- `specs/001-backend-mvp/plan.md`
- `specs/001-backend-mvp/data-model.md`
- `specs/001-backend-mvp/contracts/backend-api.openapi.yaml`

### PR DESCRIPTION
**Pull request by leehyowon14**
## 요약

- 이 PR은 `#4 Task 1.4` 범위에서 교수 등록, 목록 조회, 상세 조회, 기본 호감도 초기화까지의 코어를 구현합니다.
- 이전에는 교수와 호감도 도메인, 저장소, API, 테스트 기반이 없어 사용자별 교수 등록과 현재 호감도 조회를 검증할 수 없었습니다. 이번 변경으로 인증 사용자 기준 교수/호감도 데이터가 함께 생성되고, 사진 유무에 따른 초기 캐릭터 에셋 상태가 계약 문서와 일치하도록 고정됩니다.
- `#3` 인증 구현은 섞지 않고 `#2` 공통 기반 위에서 `professor` 모듈만 추가했습니다.

## 연결된 이슈

- Closes #4
- 원칙: 한 이슈에 대한 작업이 끝나면 해당 이슈 기준으로 바로 PR을 생성한다.

## 작업 영역

- [ ] Front-end
- [x] Back-end

## 브랜치 / 배포 대상

- 소스 브랜치: `feat/back-end/4-professor-affection-registration-core`
- 대상 브랜치: `main`
- `issueNum`은 GitHub 이슈 번호 `#123`의 숫자 부분인 `123`을 의미한다.

## 변경 내용

- `professor` 도메인과 저장소를 추가해 교수와 호감도 데이터를 PostgreSQL에 저장할 수 있게 했습니다.
- 교수 등록 시 현재 인증 사용자 기준으로 `Professor`와 `Affection(0)`를 함께 생성하고, 목록/상세 조회도 같은 사용자 스코프로만 보이도록 제한했습니다.
- `sourcePhotoUrl` 유무에 따라 `characterAssetStatus`를 `pending` 또는 `ready`로 초기화하고, `isDefaultCharacterAssets`를 함께 고정했습니다.
- `V2__create_professor_affection_tables.sql`을 추가하고 호감도 점수 범위를 `0..100`으로 제한했습니다.
- 계약, 통합, 단위 테스트를 추가해 `gender` 검증 실패, 기본 호감도 생성, 인증 사용자 스코프, 미인증 401 응답을 검증했습니다.

## 테스트

- [ ] 구현 전에 실패하는 테스트를 먼저 추가했다.
- [ ] Front-end 테스트를 추가하거나 갱신했다.
- [x] Back-end 테스트를 추가하거나 갱신했다.
- [ ] CI가 통과했다.

실행한 검증:

- `env JAVA_HOME=/opt/homebrew/opt/openjdk@21 PATH=/opt/homebrew/opt/openjdk@21/bin:/usr/bin:/bin:/usr/sbin:/sbin ./gradlew unitTest --tests com.animalleague.april.unit.professor.AffectionInitializationUnitTest`
- `env JAVA_HOME=/opt/homebrew/opt/openjdk@21 PATH=/opt/homebrew/opt/openjdk@21/bin:/usr/bin:/bin:/usr/sbin:/sbin ./gradlew contractTest --tests com.animalleague.april.contract.professor.ProfessorContractTest`
- `env JAVA_HOME=/opt/homebrew/opt/openjdk@21 PATH=/opt/homebrew/opt/openjdk@21/bin:/usr/bin:/bin:/usr/sbin:/sbin ./gradlew integrationTest --tests com.animalleague.april.integration.professor.ProfessorRegistrationIntegrationTest`

## 문서

- [ ] `Docs/`의 버전 문서를 갱신했다.
- [ ] 필요 시 API 계약을 갱신했다.
- [ ] 교차 팀 차단 이슈가 있었다면 `FE-comment` / `BE-comment` 문서를 갱신했다.
- [x] 문서 영향이 없다.

## 리뷰 메모

- 중점적으로 봐야 할 영역: `SecurityContext`에서 현재 사용자 식별자를 추출하는 fallback, 사용자 스코프 조회, `sourcePhotoUrl` 분기에 따른 초기 상태 결정
- 알려진 후속 작업: `#5`에서 실제 캐릭터 에셋 생성 파이프라인과 `characterAssets`, `representativeAssetUrl` 연동이 이어집니다.
- 이 PR이 단일 이슈 완료 단위인지, 아니라면 왜 분리하지 않았는지: `#4 Task 1.4` 범위만 포함한 단일 이슈 완료 단위입니다.
- 리뷰 코멘트 응답 규칙: 지적사항을 반영했으면 반영한 내용을, 반영하지 않았으면 반영하지 않은 이유와 대안을 각 리뷰 코멘트 답글에 남긴다.

## Summary by Sourcery

Introduce professor registration and affection core domain and APIs scoped to the authenticated user, including persistence, security-based user resolution, and audit configuration.

New Features:
- Add professor and affection domain models with PostgreSQL migrations to store per-user professors and their affection scores.
- Expose REST endpoints to create professors, list a user’s professors, and fetch detailed professor data including affection and character asset metadata.
- Derive the current user from the Spring Security context via a pluggable provider, using username-based UUIDs as a fallback identifier.

Enhancements:
- Configure JPA auditing to use an OffsetDateTime-based DateTimeProvider for consistent timestamp handling across entities.

Tests:
- Add contract tests for professor APIs covering creation, listing, detail retrieval, and validation errors for invalid gender values.
- Add integration tests verifying affection initialization at zero, character asset initial state based on source photo presence, user-scoped access to professor data, and 401 responses for unauthenticated access.
- Add unit tests ensuring professor creation initializes affection for the authenticated user and enforces affection score bounds.

### CODE REVIEW

**Comment by leehyowon14** [comment]
@codex review this pr

File: Backend/src/main/java/com/animalleague/april/professor/application/ProfessorService.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+            Gender.fromValue(request.gender()),
+            PersonalityType.fromValue(request.personalityType()),
+            request.sourcePhotoUrl()
+        );
+
+        Professor savedProfessor = professorRepository.save(professor);
+        professorRepository.flush();
+
+        UUID professorId = savedProfessor.getId();
+        if (professorId == null) {
```
[the remark this answers is not in the document]
Comment (leehyowon14) [reply Backend/src/main/java/com/animalleague/april/professor/application/ProfessorService.java]: 반영했습니다. `Professor` ID가 애플리케이션 UUID로 선할당되는 구조라 `save()` 직후 `flush()`와 null 체크는 불필요하다고 판단했고 제거했습니다. 이후 `savedProfessor.getId()`를 바로 사용하도록 정리했습니다. 관련 커밋: `d06415a`.

File: Backend/src/main/java/com/animalleague/april/professor/application/SecurityContextCurrentUserProvider.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        if (principal instanceof CharSequence sequence) {
+            try {
+                return UUID.fromString(sequence.toString());
+            } catch (IllegalArgumentException ignored) {
+                return null;
+            }
+        }
+
+        for (String methodName : List.of("id", "getId")) {
+            UUID resolved = invokeUuidAccessor(principal, methodName);
```
[the remark this answers is not in the document]
Comment (leehyowon14) [reply Backend/src/main/java/com/animalleague/april/professor/application/SecurityContextCurrentUserProvider.java:64]: 반영했습니다. 다만 비공개 accessor/field까지 reflection 범위를 넓히지는 않았습니다. principal 계약을 넓히기보다, 현재 지원 범위(UUID principal, UUID 문자열, public `id/getId`)를 코드 주석과 전용 단위 테스트로 고정하고, fallback 및 accessor 실패 경로에 debug 로그를 추가하는 쪽이 더 안전하다고 판단했습니다. 관련 커밋: `23c885d`, `b67da4d`.

File: Backend/src/test/java/com/animalleague/april/unit/professor/AffectionInitializationUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import com.animalleague.april.professor.application.ProfessorService;
+import com.animalleague.april.professor.domain.Affection;
+import com.animalleague.april.professor.domain.Professor;
+import com.animalleague.april.professor.infrastructure.AffectionRepository;
+import com.animalleague.april.professor.infrastructure.ProfessorRepository;
+
+class AffectionInitializationUnitTest {
+
+    @Test
+    void createProfessorSavesProfessorAndAffectionForAuthenticatedUser() {
```
[the remark this answers is not in the document]
Comment (leehyowon14) [reply Backend/src/test/java/com/animalleague/april/unit/professor/AffectionInitializationUnitTest.java:32]: 반영했습니다. `CurrentUserProvider`가 `null`을 반환할 때 `ProfessorService.listProfessors()`가 `401 UNAUTHORIZED`를 던지는 단위 테스트를 추가했습니다. 통합 테스트 의존 없이 서비스 레벨 인증 분기를 직접 고정하려는 의도입니다. 관련 커밋: `cc6a790`.

File: Backend/src/test/java/com/animalleague/april/unit/professor/AffectionInitializationUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        given(persistedProfessor.getGender()).willReturn(Gender.MALE);
+        given(persistedProfessor.getPersonalityType()).willReturn(PersonalityType.GENTLE);
+        given(persistedProfessor.getSourcePhotoUrl()).willReturn("https://cdn.example.com/source/prof_3.jpg");
+        given(persistedProfessor.getCharacterAssetStatus()).willReturn(CharacterAssetStatus.PENDING);
+        given(persistedProfessor.getRepresentativeAssetUrl()).willReturn(null);
+        given(persistedProfessor.isDefaultCharacterAssets()).willReturn(false);
+        given(professorRepository.save(any(Professor.class))).willReturn(persistedProfessor);
+        given(affectionRepository.save(any(Affection.class))).willAnswer(invocation -> invocation.getArgument(0));
+
+        ProfessorCreateResponse response = professorService.createProfessor(
```
[the remark this answers is not in the document]
Comment (leehyowon14) [reply Backend/src/test/java/com/animalleague/april/unit/professor/AffectionInitializationUnitTest.java:59]: 반영했습니다. `SecurityContextCurrentUserProvider` 전용 단위 테스트를 추가해서 UUID principal, public `getId()` accessor, 이름 기반 fallback, anonymous 분기를 각각 검증했습니다. principal 해석 규칙이 바뀔 때 빠르게 깨지도록 범위를 명시했습니다. 관련 커밋: `b67da4d`.

File: Backend/src/test/java/com/animalleague/april/contract/professor/ProfessorContractTest.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        mockMvc.perform(get("/api/professors/11111111-1111-1111-1111-111111111111"))
+            .andExpect(status().isOk())
+            .andExpect(jsonPath("$.professor.id").value("11111111-1111-1111-1111-111111111111"))
+            .andExpect(jsonPath("$.affection.professorId").value("11111111-1111-1111-1111-111111111111"))
+            .andExpect(jsonPath("$.affection.affectionScore").value(0))
+            .andExpect(jsonPath("$.characterAssets").value(org.hamcrest.Matchers.hasSize(0)));
+    }
+
+    @Test
+    void invalidGenderReturns400() throws Exception {
```
[the remark this answers is not in the document]
Comment (leehyowon14) [reply Backend/src/test/java/com/animalleague/april/contract/professor/ProfessorContractTest.java:197]: 반영했습니다. 기존 `invalid gender` 외에 `blank professorName`, `blank gender`, `blank personalityType`, `invalid personalityType` 경로를 contract test로 추가했습니다. 생성 요청 validation 범위를 문서화하는 성격의 보강입니다. 관련 커밋: `b47adf2`.

File: Backend/src/test/java/com/animalleague/april/contract/professor/ProfessorContractTest.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+            .andExpect(jsonPath("$.professor.id").value("11111111-1111-1111-1111-111111111111"))
+            .andExpect(jsonPath("$.professor.professorName").value("홍길동"))
+            .andExpect(jsonPath("$.professor.gender").value("male"))
+            .andExpect(jsonPath("$.professor.personalityType").value("gentle"))
+            .andExpect(jsonPath("$.professor.characterAssetStatus").value("ready"))
+            .andExpect(jsonPath("$.professor.isDefaultCharacterAssets").value(true));
+    }
+
+    @Test
+    void createProfessorWithSourcePhotoReturnsPendingProfessorEnvelope() throws Exception {
```
[the remark this answers is not in the document]
Comment (leehyowon14) [reply Backend/src/test/java/com/animalleague/april/contract/professor/ProfessorContractTest.java:78]: 반영했습니다. whitespace-only `sourcePhotoUrl` 입력에 대해 응답 계약이 no-photo 케이스(`ready`, 기본 에셋 사용, `sourcePhotoUrl = null`)로 보이도록 contract test를 추가했습니다. 저장 시 정규화 자체는 통합 테스트에서 별도로 확인했습니다. 관련 커밋: `29f342e`.

File: Backend/src/test/java/com/animalleague/april/integration/professor/ProfessorRegistrationIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+        assertThat(professor.getCharacterAssetStatus()).isEqualTo(CharacterAssetStatus.READY);
+        assertThat(professor.isDefaultCharacterAssets()).isTrue();
+
+        Affection affection = affectionRepository.findByProfessorIdAndUserId(professor.getId(), userId)
+            .orElseThrow();
+        assertThat(affection.getAffectionScore()).isZero();
+    }
+
+    @Test
+    void createProfessorWithSourcePhotoStartsInPendingState() throws Exception {
```
[the remark this answers is not in the document]
Comment (leehyowon14) [reply Backend/src/test/java/com/animalleague/april/integration/professor/ProfessorRegistrationIntegrationTest.java:74]: 반영했습니다. whitespace-only `sourcePhotoUrl`가 실제로 `null`로 저장되고, `READY`, `isDefaultCharacterAssets = true`, 초기 호감도 `0`으로 이어지는지 integration test를 추가했습니다. 관련 커밋: `e920430`.

File: Backend/src/test/java/com/animalleague/april/integration/professor/ProfessorRegistrationIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+            .andExpect(status().isCreated());
+
+        Professor professor = professorRepository.findAllByUserIdOrderByCreatedAtDesc(userId).getFirst();
+        assertThat(professor.getSourcePhotoUrl()).isEqualTo("https://cdn.example.com/source/prof_2.jpg");
+        assertThat(professor.getCharacterAssetStatus()).isEqualTo(CharacterAssetStatus.PENDING);
+        assertThat(professor.isDefaultCharacterAssets()).isFalse();
+    }
+
+    @Test
+    void professorListAndDetailAreScopedToAuthenticatedUser() throws Exception {
```
[the remark this answers is not in the document]
Comment (leehyowon14) [reply Backend/src/test/java/com/animalleague/april/integration/professor/ProfessorRegistrationIntegrationTest.java:128]: 반영했습니다. 목록 응답에서 `alice`의 교수만 보이는지 `hasSize(1)`와 Bob 교수의 id/name이 포함되지 않는 조건까지 추가했습니다. 상세 404 외에 목록 누출 회귀도 잡기 위한 보강입니다. 관련 커밋: `6f78c87`.

File: Backend/src/test/java/com/animalleague/april/integration/professor/ProfessorRegistrationIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+
+        mockMvc.perform(
+                get("/api/professors/{professorId}", bobProfessor.getId())
+                    .with(SecurityMockMvcRequestPostProcessors.user("alice"))
+            )
+            .andExpect(status().isNotFound());
+    }
+
+    @Test
+    void unauthenticatedProfessorRequestReturns401() throws Exception {
```
[the remark this answers is not in the document]
Comment (leehyowon14) [reply Backend/src/test/java/com/animalleague/april/integration/professor/ProfessorRegistrationIntegrationTest.java:164]: 반영했습니다. unauthenticated `POST /api/professors`에 대해 `401`과 `UNAUTHORIZED` 코드를 검증하는 integration test를 추가했습니다. 조회와 생성 모두 같은 인증 경계를 갖도록 맞춘 변경입니다. 관련 커밋: `2720447`.

File: Backend/src/test/java/com/animalleague/april/integration/support/PostgresIntegrationTest.java [hunk 10]
```diff
@@ -17,6 +18,7 @@
 @SpringBootTest
 @ActiveProfiles("test")
 @Testcontainers(disabledWithoutDocker = true)
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_CLASS)
```
[the remark this answers is not in the document]
Comment (leehyowon14) [reply Backend/src/test/java/com/animalleague/april/integration/support/PostgresIntegrationTest.java]: 반영했습니다. 말씀하신 대로 base class의 전역 `@DirtiesContext`는 제거하고, 실제로 컨텍스트 오염 이슈를 만들었던 `AuthFlowIntegrationTest`와 `ProfessorRegistrationIntegrationTest`에만 범위를 좁혔습니다. 이후 Linux Docker에서 전체 `integrationTest`를 다시 돌려 회귀가 없는 것도 확인했습니다. 관련 커밋: `d718963`.

File: Backend/src/main/java/com/animalleague/april/professor/api/ProfessorCreateRequest.java [hunk 11]
```diff
@@ -0,0 +1,20 @@
+package com.animalleague.april.professor.api;
+
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Pattern;
+
+public record ProfessorCreateRequest(
+    @NotBlank(message = "교수명은 비어 있을 수 없습니다.")
+    String professorName,
```
[the remark this answers is not in the document]
Comment (leehyowon14) [reply Backend/src/main/java/com/animalleague/april/professor/api/ProfessorCreateRequest.java:10]: 반영했습니다. `professorName`에 `@Size(max = 100)`를 추가하고, 101자 입력이 `400 VALIDATION_ERROR`로 내려가는 contract test를 함께 추가했습니다. DB 제약으로 늦게 터지던 케이스를 요청 검증 단계로 끌어올린 변경입니다. 관련 커밋: `781787a`.

**Comment by leehyowon14** [comment]
상단 review body 코멘트에 대한 처리 현황입니다.

- `SecurityContextCurrentUserProvider`의 fallback 진단성은 반영했습니다. 지원 principal 형태를 코드/테스트로 고정했고, fallback 및 accessor 실패 경로에 debug 로그를 추가했습니다. 관련 커밋: `23c885d`, `b67da4d`
- `ProfessorCreateRequest`의 enum 문자열을 annotation에서 직접 유도하는 변경은 이번 PR에는 반영하지 않았습니다. Bean Validation의 `@Pattern`은 compile-time constant가 필요해서, 이를 깔끔하게 해결하려면 custom validator나 별도 제약 애노테이션 도입이 필요합니다. 현재 PR 범위를 넘는다고 판단했고, 대신 validation contract coverage는 확장했습니다. 관련 커밋: `b47adf2`, `781787a`
